File: BatteryHealthApp/app/src/main/java/com/batteryhealth/app/BatteryHealthResult.java
package com.batteryhealth.app;

import java.util.List;
import java.util.ArrayList;

/**
 * 电池健康度结果数据类
 */
public class BatteryHealthResult {
    // 基础结果
    public float healthPercentage;     // 健康度百分比 (0-100)
    public String grade;               // 等级: A+, A, B, C, D, F
    public String gradeColor;          // 等级颜色
    public String gradeDescription;    // 等级描述

    // 详细信息
    public String diagnosisText;       // 诊断文字
    public List<String> suggestions;   // 使用建议
    public HealthFactors factors;      // 健康因子

    // 置信度 (0=NONE, 1=LOW, 2=MEDIUM, 3=HIGH)
    public int confidence;

    // 辅助数据
    public float estimatedResistanceMohm;  // 估算内阻 (mΩ)
    public int remainingLifespanMonths;    // 预估剩余寿命 (月)

    public BatteryHealthResult() {
        healthPercentage = 0f;
        grade = "F";
        gradeColor = "#9E9E9E";
        gradeDescription = "未知状态";
        diagnosisText = "";
        suggestions = new ArrayList<>();
        factors = new HealthFactors();
        confidence = 0;
        estimatedResistanceMohm = 0f;
        remainingLifespanMonths = 0;
    }

    // 获取置信度描述
    public String getConfidenceText() {
        switch (confidence) {
            case 3: return "高";
            case 2: return "中";
            case 1: return "低";
            default: return "无";
        }
    }

    // 获取健康度百分比描述
    public String getHealthPercentageText() {
        return String.format("%.1f%%", healthPercentage);
    }

    // 获取等级完整描述
    public String getGradeFullText() {
        return grade + " - " + gradeDescription;
    }

    // 获取剩余寿命描述
    public String getRemainingLifespanText() {
        if (remainingLifespanMonths > 0) {
            return "约 " + remainingLifespanMonths + " 个月";
        }
        return "无法估算";
    }

    // 获取内阻描述
    public String getResistanceText() {
        if (estimatedResistanceMohm > 0) {
            return String.format("%.2f mΩ", estimatedResistanceMohm);
        }
        return "未估算";
    }

    // 是否有有效数据
    public boolean hasValidResult() {
        return healthPercentage > 0 || (factors != null && factors.availableFactors > 0);
    }

    // 获取因子详情描述
    public String getFactorsDetailText() {
        StringBuilder sb = new StringBuilder();
        if (factors != null) {
            if (factors.capacityRetention > 0) {
                sb.append("容量保持率: ").append(String.format("%.1f%%", factors.capacityRetention)).append("\n");
            }
            if (factors.cycleDecay > 0) {
                sb.append("循环衰减: ").append(String.format("%.1f%%", factors.cycleDecay)).append("\n");
            }
            if (factors.resistanceGrowth > 0) {
                sb.append("内阻增长: ").append(String.format("%.1f%%", factors.resistanceGrowth)).append("\n");
            }
            if (factors.thermalAging > 0) {
                sb.append("温度老化: ").append(String.format("%.1f%%", factors.thermalAging)).append("\n");
            }
            if (factors.chargingDamage > 0) {
                sb.append("充电损伤: ").append(String.format("%.1f%%", factors.chargingDamage)).append("\n");
            }
        }
        return sb.toString();
    }
}

File: digiguide/android/app/src/main/java/com/digiguide/model/BatteryHealthResult.kt
package com.digiguide.model

/**
 * 电池健康度结果
 */
data class BatteryHealthResult(
    var healthPercentage: Float = 0f,
    var grade: String = "F",
    var capacityRetention: Float? = null,
    var cycleDecay: Float? = null,
    var resistanceGrowth: Float? = null,
    var thermalAging: Float? = null,
    var chargingDamage: Float? = null,
    var diagnosisText: String = "",
    var suggestions: List<String> = emptyList(),
    var estimatedResistanceMohm: Float? = null,
    var remainingLifespanMonths: Int? = null,
    var confidence: ConfidenceLevel = ConfidenceLevel.NONE,
    // 循环次数专项分析
    var cycleCount: Int? = null,
    var cycleGrade: String = "未知",
    var cyclePercentUsed: Float = 0f,
    var estimatedRemainingCycles: Int? = null
) {
    /**
     * 置信度级别
     */
    enum class ConfidenceLevel {
        HIGH,    // 高置信度（>=4个因子）
        MEDIUM,  // 中置信度（>=2个因子）
        LOW,     // 低置信度（>=1个因子）
        NONE     // 无置信度（无可用因子）
    }

    /**
     * 获取等级颜色
     */
    fun getGradeColor(): String {
        return when (grade) {
            "A+" -> "#4CAF50"
            "A" -> "#8BC34A"
            "B" -> "#FFC107"
            "C" -> "#FF9800"
            "D" -> "#F44336"
            "F" -> "#9C27B0"
            else -> "#9E9E9E"
        }
    }

    /**
     * 获取等级描述
     */
    fun getGradeDescription(): String {
        return when (grade) {
            "A+" -> "电池状态极佳，几乎无老化"
            "A" -> "电池状态良好，轻微老化"
            "B" -> "电池状态一般，中度老化"
            "C" -> "电池状态较差，明显老化"
            "D" -> "电池状态很差，严重老化"
            "F" -> "电池状态极差，建议更换"
            else -> "未知状态"
        }
    }

    /**
     * 是否需要更换电池
     */
    fun needsReplacement(): Boolean = healthPercentage < 60f

    /**
     * 是否处于警告状态
     */
    fun isWarningState(): Boolean = healthPercentage < 80f

    /**
     * 是否处于良好状态
     */
    fun isGoodState(): Boolean = healthPercentage >= 90f

    /**
     * 获取可用因子数量
     */
    fun getAvailableFactorsCount(): Int {
        var count = 0
        if (capacityRetention != null) count++
        if (cycleDecay != null) count++
        if (resistanceGrowth != null) count++
        if (thermalAging != null) count++
        if (chargingDamage != null) count++
        return count
    }

    /**
     * 获取详细诊断报告
     */
    fun getDetailedReport(): String {
        val sb = StringBuilder()
        sb.append("=== 电池健康度分析报告 ===\n\n")
        sb.append("综合健康度: ${healthPercentage.toInt()}% ($grade)\n")
        sb.append("状态描述: ${getGradeDescription()}\n")
        sb.append("置信度: ${confidence.name} (${getAvailableFactorsCount()}个因子)\n\n")

        sb.append("=== 详细分析 ===\n")
        if (capacityRetention != null) {
            sb.append("容量保持率: ${(capacityRetention!! * 100).toInt()}%\n")
        }
        if (cycleDecay != null) {
            sb.append("循环衰减因子: ${(cycleDecay!! * 100).toInt()}%\n")
        }
        if (resistanceGrowth != null) {
            sb.append("内阻增长因子: ${(resistanceGrowth!! * 100).toInt()}%\n")
        }
        if (thermalAging != null) {
            sb.append("温度老化因子: ${(thermalAging!! * 100).toInt()}%\n")
        }
        if (chargingDamage != null) {
            sb.append("充电损伤因子: ${(chargingDamage!! * 100).toInt()}%\n")
        }

        if (estimatedResistanceMohm != null) {
            sb.append("\n估算内阻: ${estimatedResistanceMohm!!.toInt()}mΩ\n")
        }
        if (remainingLifespanMonths != null) {
            sb.append("预估剩余寿命: ${remainingLifespanMonths!!}个月\n")
        }

        if (suggestions.isNotEmpty()) {
            sb.append("\n=== 使用建议 ===\n")
            for (suggestion in suggestions) {
                sb.append("- $suggestion\n")
            }
        }

        return sb.toString()
    }
}